### DISCUSSION TITLE
How to get list of available devices using Silk.NET.OpenAL?

### DISCUSSION THREAD

**Discussion by asheesh1996** [dotnet/Silk.NET#573] [Q&A] [answered]
I want to detect microphone and speakers connected to my windows running dotnet5, I went through this [example](https://github.com/dotnet/Silk.NET/tree/main/examples/CSharp/OpenAL%20Demos/WavePlayer) but it uses default connected device, I want to enumerate though list of devices and play the audio if any examples are available please suggest or any API documentation for getting list of available connected devices?

**Comment by Perksey** [comment, marked as answer]
Ah apologies, I had an answer written for this the day you opened this but apparently never hit send.

You'll have to use ALContext and its enumaration extensions (which are not guaranteed to be available as they are extensions!)

These are in the Silk.NET.OpenAL.Extensions.Enumeration package. Using this package, you can get a string list containing all available device specifiers. It'll look something like this:
```cs
using var alc = ALContext.GetApi();
string deviceSpecifier = "";
if (alc.IsExtensionPresent(null, "ALC_ENUMERATION_EXT"))
{
    using var enumeration = alc.GetExtension<Enumeration>();
    foreach (var device in enumeration.GetStringList(null, GetEnumerationContextStringList.DeviceSpecifiers))
    {
        Console.WriteLine($"Got device \"{device}\"");
        deviceSpecifier = device;
    }
}

var device = alc.OpenDevice(deviceSpecifier);
// ...
```

**Reply by Perksey** [reply]
Note I appreciate that the whole IsExtensionPresent/GetExtension thing is ugly. These are handwritten bindings and don't have the unified `TryGetExtension` that our other bindings do - we have an up-for-grabs issue #574 for fixing this.

**Reply by LaXiS96** [reply]
The code you provided does not compile/run under v2.7.0.
There is no overload for `GetExtension<>()` that takes zero arguments, only one that takes a `Device*` argument.
Even passing `null` for the `Device*` argument only results in a runtime exception on the same line:
```
An unhandled exception of type 'System.MissingMethodException' occurred in System.Private.CoreLib.dll
Constructor on type 'Silk.NET.OpenAL.Extensions.Enumeration.Enumeration' not found.
```
This was the test code:
```csharp
using System;
using Silk.NET.OpenAL;
using Silk.NET.OpenAL.Extensions.Enumeration;

namespace ConsoleApp1
{
    class Program
    {
        static unsafe void Main(string[] args)
        {
            using var alc = ALContext.GetApi(true);

            if (alc.IsExtensionPresent(null, "ALC_ENUMERATION_EXT"))
            {
                using var enumeration = alc.GetExtension<Enumeration>(null);
                foreach (var device in enumeration.GetStringList(GetEnumerationContextStringList.DeviceSpecifiers))
                {
                    Console.WriteLine(device);
                }
            }
        }
    }
}
```
Installed NuGet packages:
- Silk.NET.OpenAL 2.7.0
- Silk.NET.OpenAL.Extensions.Enumeration 2.7.0
- Silk.NET.OpenAL.Soft.Native 1.21.1.1

**Reply by Perksey** [reply]
In #598 and #602 we fix the behaviour you're encountering by:
- adding a `TryGetExtension` method to eliminate the error-prone `IsExtensionPresent`/`GetExtension` pair
- fixing a lot of extension-related bugs

This will be in the 2.8 update which will be released as soon as a couple of PRs are merged.
